### DISCUSSION TITLE
RDF4J SHACL Implementation and W3C Awesome Semantic Shapes list

### DISCUSSION THREAD

**Discussion by amivanoff** [eclipse-rdf4j/rdf4j#5137] [General]
Hello!

We are gathering an initial version the [w3c-cg/awesome-semantic-shapes](https://github.com/w3c-cg/awesome-semantic-shapes) list. It's not published yet. "First pull requests" stage.

Current entry format for the SHACL Validators list is like this one:

[link_to_the_docs_or_repostory] activity_bage - Supports parts of SHACL spec; optional other worth mentioning; [optional-playground-link]; `LICENSE` `ProgrammingLanguage`.

Examples:

- [Jena SHACL](https://jena.apache.org/documentation/shacl/index.html) <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/apache/jena" align="top"> - Supports SHACL Core, SHACL-SPARQL; `Apache-2.0` `Java`.
- [RDF4J SHACL Sail](https://rdf4j.org/documentation/programming/shacl) <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/eclipse-rdf4j/rdf4j" align="top"> - Supports XXX; `BSD-3-Clause` `Java`.
- [TopBraid SHACL API](https://github.com/TopQuadrant/shacl) <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/TopQuadrant/shacl" align="top"> - Supports SHACL Core, SHACL-SPARQL, SHACL rules; based on Jena; `BSD-3-Clause` `Java`.
- [TopBraid SHACL API Extended](https://github.com/SHACL-X/shacl-x) <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/SHACL-X/shacl-x" align="top"> - Supports SHACL Core, SHACL-SPARQL, SHACL rules, SHACL-JS; fork of the TopBraid SHACL API with added SHACL-JS, based on GraalVM Polyglot; `BSD-3-Clause` `Java`.
- [rdf-validate-shacl](https://github.com/zazuko/rdf-validate-shacl) <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/zazuko/rdf-validate-shacl" align="top"> - Supports SHACL Core; pure JavaScript validator on top of the [RDFJS](https://rdf.js.org/) stack; [playground](https://zazuko.github.io/shacl-playground); `MIT` `JavaScript`.

The questions are:
- What is the better name of the RDF4J Shacl implementation? RDF4J SHACL? RDF4J SHACL Sail? RDF4J SHACL Engine (like in the documentation)?
- What should we put into the "Supports" and "worth mentioning" sections?

As far as I know, RDF4J Shacl does not participate in [SHACL Test Suite and Implementation Report](https://w3c.github.io/data-shapes/data-shapes-test-suite/)?

**Comment by hmottestad** [comment]
RDF4J SHACL Engine might be a better name since we now have both the ShaclSail and the ShaclValidator. Our SHACL implementation isn't complete. We are missing the disjoint feature of the qualified shapes and also the oneOrMorePath, zeroOrMorePath and zeroOrOnePath. I'm currently working on those last three path implementations.

The biggest feature of our SHACL implementation is that it can be used as a wrapper (ShaclSail) around a database for incremental validation. If your database has a hundred million triples in it and you add 10 more in a transaction, then we can efficiently validate just those changes and abort the transaction if it would leave the database in an invalid state. Same is true for updating/deleting data.

**Reply by hmottestad** [reply]
And no, we haven't submitted a result to the SHACL Test Suite and Implementation Report. Though we do use the test suite as part of our unit tests.

**Reply by amivanoff** [reply]
So, the description could be something like this?

"Supports SHACL Core (without some property paths), SHACL-SPARQL"

But no SHACL rules, nor SHACL-JS?

**Reply by hmottestad** [reply]
We support SPARQL targets from SHACL Advanced and SPARQL-based Constraints from SHACL-SPARQL. This is the same as Jena. 

I would write: "Supports SHACL Core (without some property paths), SHACL-SPARQL, incremental validation"

Since AFAIK there are no other implementations that support incremental validation.
